### DISCUSSION TITLE
Any way to test WebUI changes without flashing to device (local dev debugging)

### DISCUSSION THREAD

**Discussion by prashker** [gysmo38/mitsubishi2MQTT#275] [General]
I want to try adding new features to UI but the only real unit I have is in "production" - does any IDE allow me to develop locally on my PC without a real unit? Any ideas from other people who've contributed to the project?

**Comment by floatplane** [comment]
You should be able to attach an ESP board to your machine and push changes to it via Arduino or PlatformIO. The software can run without being attached to any heat pump hardware. If you don't have a spare board lying around, then they're super-cheap on Amazon.

I have a fork of this project ([floatplane/MitsuQTT](https://github.com/floatplane/MitsuQTT)) that has diverged pretty far from this one, so this isn't a direct answer to your question but might provide inspiration if you want to make changes to this repo. One big modification I made was to split frontend views out into Mustache templates which then can be rendered by other software. In the repo I've got a [simple Ruby webserver script](https://github.com/floatplane/MitsuQTT?tab=readme-ov-file#testing-web-changes-locally) that allows rendering views locally without any dedicated hardware required. Feel free to take a look if that's helpful!

**Comment by prashker** [comment]
I was hoping I could do it without another board - I know it works without actually being connected to the unit (its how I tested some initial code changes I pushed upstream) but I suspect just spending the few bucks for another board will work.

In any case, thank you for the info - I will mark this as closed as I suspect there is no solution other than buying another (which I will do). As for your aside, thank you for the link, interesting fork, will give it a check!
